google/resource_compute_global_forwarding_rule.go

```diff
@@ -194,6 +234,31 @@ func resourceComputeGlobalForwardingRuleUpdate(d *schema.ResourceData, meta inte
 
 		d.SetPartial("target")
 	}
+	if d.HasChange("labels") {
+		// Only able to set labels via the beta API right now. Fail loudly if we're told to use a different api.
+		if computeApiVersion != v0beta {
```

Comment:
Is that even possible? The computeApiVersion should prevent this case from happening.

Reply:
It's not, at least in the current state. This is more to make sure we get a nice error message if someone changes the `Feature` to no longer be in v0beta (for example, if labels is supported in v1).

I could have written it to use the configured client api version but that felt silly since you always have to use the beta api currently.

Reply:
Doesn't guarding `computeApiVersion` like this (and above) just duplicate the information about the required version for a feature that lives in the `Feature` struct? It now lives in two places, and they might drift.

Guarding against incorrect changes to the `Feature` version for some feature is done by the tests already; they would fail if someone incorrectly changed the API version from `v0beta` to `v1`, since they would read at Beta and see that the value is not what is expected.

Reply:
+1

Reply:
The tests would catch this without the explicit guard I wrote as they wouldn't see the label being set anymore upon creation. This would probably be surprising; hence the explicit guard.

If this was a normal feature that was just another beta attribute to be set on the `GlobalForwardingRule` on creation, then everything would just work. However because only the beta api has a `SetLabels` method, we can only write the beta codepath.

Does that make sense?

Reply:
My suggestion was to remove this explicit guard since the Feature already guards it. Like you said, tests would catch a bug if this wasn't true. And leave only the beta codepath and maybe a `// Labels for GlobalForwardingRules are in beta` comment.

Reply:
Actually, I was wrong; without the guard, then we will always use the beta api to call `SetLabels`. All the tests would pass but we'd be using the wrong API.

Reply:
Which is what we want! If labels gets modified (remember, we are in a HasChange("labels") block)  then we need to be using the API to either add, update or delete labels.

Reply:
Yes, that works now. But in the future, when `Labels` is supported on v1 and someone removes it from the beta Features list, we will continue to use the beta API instead of the v1 api and all of the tests will continue to pass. This is undesirable as beta APIs have different SLAs compared to non-beta.

Reply:
Yes, once Labels becomes GA, we will need to update this code to remove it from the future and add the v1 codepath

Reply:
The 'add the v1 codepath' is something I might forget to do, especially as a normal beta feature would not require an explicit 'add the v1 codepath' (even more surprisingly, other resources that use `Labels` in beta don't require an explicit codepath to be added as they are able to set labels upon creation). I prefer the tests to catch this bug rather than relying on someone to explicitly check for this.

Can you quantify what the cost is of leaving the guard in? I'm not seeing why this is controversial.

Reply:
I see the value in your argument that 'at some point it will become GA and this safeguard will make sure we don't forget to update this part of the code'. I am fine with leaving it in.

If we didn't know for sure that we would have to update this code, I would feel strongly for removing an explicit guard guarding nothing because IMO this would be equivalent to leaving dead code.

But again, in this particular situation, we know that GA is coming for sure and I see the value in leaving it in.

Thanks

Reply:
Good point! This does help us more explicitly manage the version.

What if we added the switch on `computeApiVersion` now, and made the `v1` block throw this error instead of wrapping all this in an if? That way we still fail tests, and it's very explicit where the code will need to go to fix the problem; we'll also make smaller diffs when the feature goes `v1` since we will match the convention for versioned resources more closely.

Reply:
Done. I'm not sure it will make a smaller diff but it's pretty much isomorphic to the original code.